pylint/checkers/utils.py

```diff
@@ -1492,3 +1492,30 @@ def is_assign_name_annotated_with(node: astroid.AssignName, typing_name: str) ->
     ):
         return True
     return False
+
+
+def get_iterating_dictionary_name(
+    node: Union[astroid.For, astroid.Comprehension]
+) -> Optional[str]:
+    """Get the name of the dictionary which keys are being iterated over on
+    a `astroid.For` or `astroid.Comprehension` node.
+
+    If the iterating object is not either the keys method of a dictionary
+    or a dictionary itself, this returns None.
+    """
+    # Is it a proper keys call?
+    is_iterating_dict_keys = (
+        isinstance(node.iter, astroid.Call)
+        and isinstance(node.iter.func, astroid.Attribute)
+        and node.iter.func.attrname == "keys"
+        and isinstance(safe_infer(node.iter.func), astroid.BoundMethod)
+    )
+    # Is it a proper dictionary?
+    is_iterating_dict = isinstance(
+        node.iter, (astroid.Name, astroid.Attribute)
+    ) and isinstance(safe_infer(node.iter), astroid.Dict)
+
+    if is_iterating_dict or is_iterating_dict_keys:
+        return node.iter.as_string().partition(".keys")[0]
+
+    return None
```

Comment:
I understand the usage of rpartition here -- the new test case was written because it was failing on (left) partition :smile: 

Thank you so much for your continued patience with reviewing my code and giving suggestions -- I'm learning a lot from this! 

Let me work on the changes and ping again when I'm done